### PR TITLE
Document cartridge states

### PR DESCRIPTION
**Pull request by xuniq**
Preview: https://tarantool-develop.herokuapp.com/en/doc/gh-1538/book/cartridge/cartridge_dev/#cluster-instance-lifecycle

Fixes tarantool/doc#1538

- [x] Add state-machine.rst to /rst/topics/
- [x] Fix grammar mistakes
- [x] Minor text rewrite and update
- [x] Choose how to render .uml files

### CODE REVIEW

File: rst/conf.py [hunk 1]
```diff
@@ -25,8 +25,9 @@
 
 extensions = [
     'TarantoolSessionLexer',
+    'sphinxcontrib.plantuml',
 ]
-
+plantuml = 'java -jar /path/to/plantuml.jar'
```
Comment (NickVolynkin) [inline rst/conf.py:30]:
```suggestion
# plantuml plugin configuration
# https://github.com/sphinx-contrib/plantuml/
plantuml = 'java -jar /path/to/plantuml.jar'
plantuml_output_format = 'svg'

```

File: rst/dev/states.rst [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+Every instance in the cluster has an internal state machine. It helps to
+manage cluster operation and describe a distributed system
+simpler.
+
+..  uml::  ../doc/rst/uml/state-machine.uml
+
+
+Instance lifecycle starts from ``cartridge.cfg`` call. During initialization
+cartridge instance binds TCP (iproto) and UDP sockets
```
Comment (alexandra-mara) [inline rst/dev/states.rst:14]:
```suggestion
Cartridge instance binds TCP (iproto) and UDP sockets
```

**Review by rosik — COMMENTED** [review]
We discussed it orally.

Plantuml is a heavyweight dependency that depends on java, so I'd like it to be optional. To do so, let's reproduce the pattern used for fetching GraphQL schema.

1. The resulting files (that is, `schema.graphql` and SVG images in the future) are committed in the repo.
2. CI checks that those files are up to date.
  - GraphQL schema is always regenerated and compared to the existing file, see https://github.com/tarantool/cartridge/blob/master/fetch-schema.sh
  - PlantUML images could be compared by hash.
3. We should provide an easy way to re-generate those files and describe in the [contribution guide](https://github.com/tarantool/cartridge/blob/master/CONTRIBUTING.rst)

File: rst/topics/state-machine.rst [hunk 3]
```diff
@@ -86,9 +85,9 @@ Configuring arguments for ``box.cfg`` if snapshots or config files are
 not present. ``box.cfg`` execution. Setting up users and stopping
 ``remote-control``. The instance will try to start listening to full-featured
 iproto protocol. In case of failed attempt instance will change its
-state to ``BootError``. If there is no replicaset in cluster-wide
-config, the instance will set the state to ``BootError``. If
-everything is ok, the instance is set to ``ConnectingFullmesh``.
+state to ``BootError``. On success, the instance enters the ``ConnectingFullmesh`` state.
+If there is no replicaset in cluster-wide
+config, the instance will set the state to ``BootError``.
```
Comment (alexandra-mara) [inline rst/topics/state-machine.rst:90]: А тут теперь вышло сначала описание BootError, потом ConnectingFullmesh и снова BootError, это ок? Раньше ошибки были вместе сгруппированы

File: rst/topics/state-machine.rst [hunk 4]
```diff
@@ -118,8 +117,8 @@ ConnectingFullmesh
 
 
 During this state, a configuration of servers and replicasets is being
-performed. Eventually, cluster topology, which is described in config, is
-implemented. But in case of an error instance the state is changed to
+performed. Eventually, cluster topology, which is described in the config, is
+implemented. But in case of an error instance, the state moves to
```
Comment (alexandra-mara) [inline rst/topics/state-machine.rst:121]: В этой строчке либо instance лишнее, либо что-то другое надо поправить

**Comment by akudiyar** [comment]
You may avoid adding a Java dependency by deploying a plantUML Web server as an internal service (see https://github.com/plantuml/plantuml-server). Also running diagram generation in Docker is an option too.

**Review by rosik — APPROVED** [review]
This PR needs a follow-up with minor edits, but it's already big enough. So I'm going to merge it and make edits in a new PR.

**Pull request merged**